x-pack/plugins/observability/server/lib/annotations/create_annotations_client.ts

```diff
@@ -0,0 +1,187 @@
+
+interface IndexDocumentResponse {
+  _shards: {
+    total: number;
+    failed: number;
+    successful: number;
+  };
+  _index: string;
+  _type: string;
+  _id: string;
+  _version: number;
+  _seq_no: number;
+  _primary_term: number;
+  result: string;
+}
+
+export function createAnnotationsClient(params: {
+  index: string;
+  apiCaller: APICaller;
+  logger: Logger;
+}) {
+  const { index, apiCaller, logger } = params;
+
+  const initIndex = () =>
+    createOrUpdateIndex({
+      index,
+      mappings: {
+        dynamic: 'strict',
+        properties: {
+          annotation: {
+            properties: {
+              type: {
+                type: 'keyword',
+              },
+            },
+          },
+          message: {
+            type: 'text',
+          },
+          tags: {
+            type: 'keyword',
+          },
+          '@timestamp': {
+            type: 'date',
+          },
+          event: {
+            properties: {
+              created: {
+                type: 'date',
+              },
+            },
+          },
+          service: {
+            properties: {
+              name: {
+                type: 'keyword',
+              },
+              environment: {
+                type: 'keyword',
+              },
+              version: {
+                type: 'keyword',
+              },
+            },
+          },
+        },
+      },
+      apiCaller,
+      logger,
+    });
+
+  return {
+    search: async (searchParams: SearchParams): Promise<SearchResponse<Annotation>> => {
+      const { start, end, size, annotation, tags, filter } = searchParams;
+
+      try {
+        return await apiCaller('search', {
+          index,
+          body: {
+            size: size ?? 50,
+            query: {
+              bool: {
+                filter: [
+                  {
+                    range: {
+                      '@timestamp': {
+                        gte: start,
+                        lt: end,
+                      },
+                    },
+                  },
+                  ...(annotation?.type ? [{ term: { 'annotation.type': annotation.type } }] : []),
+                  ...(tags ? [{ terms: { tags } }] : []),
+                  ...(filter ? [filter] : []),
+                ],
+              },
+            },
+          },
+        });
+      } catch (err) {
+        if (err.body?.error?.type !== 'index_not_found_exception') {
+          throw err;
+        }
+        return {
+          hits: {
+            total: 0,
+            max_score: 0,
+            hits: [],
+          },
+          took: 0,
+          _shards: {
+            failed: 0,
+            skipped: 0,
+            successful: 0,
+            total: 0,
+          },
+          timed_out: false,
+        };
+      }
+    },
+    create: async (
+      createParams: CreateParams
+    ): Promise<{ _id: string; _index: string; _source: Annotation }> => {
+      await initIndex();
```

Comment:
Could putting the mappings on every create have unintended consequences?

Reply:
I'll add a `indices.exists` call before.